File: documentation/manual/working/scalaGuide/main/dependencyinjection/code/CompileTimeDependencyInjection.scala
/*
 * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
 */

package scalaguide.dependencyinjection

import java.io.File

import _root_.controllers.AssetsMetadata
import org.specs2.mutable.Specification

class CompileTimeDependencyInjection extends Specification {
  import play.api._

  val environment = Environment(new File("."), this.getClass.getClassLoader, Mode.Test)

  "compile time dependency injection" should {
    "allow creating an application with the built in components from context" in {
      val context = ApplicationLoader.Context
        .create(environment, Map("play.application.loader" -> classOf[basic.MyApplicationLoader].getName))
      val components  = new messages.MyComponents(context)
      val application = ApplicationLoader(context).load(context)
      application must beAnInstanceOf[Application]
      components.router.documentation must beEmpty
    }
    "allow using other components" in {
      val context    = ApplicationLoader.Context.create(environment)
      val components = new messages.MyComponents(context)
      components.application must beAnInstanceOf[Application]
      components.myComponent must beAnInstanceOf[messages.MyComponent]
    }
    "allow declaring a custom router" in {
      val context    = ApplicationLoader.Context.create(environment)
      val components = new routers.MyComponents(context)
      components.application must beAnInstanceOf[Application]
      components.router must beAnInstanceOf[scalaguide.dependencyinjection.Routes]
    }
  }
}

package basic {
//#basic
  import play.api._
  import play.api.routing.Router
  import play.api.ApplicationLoader.Context
  import play.filters.HttpFiltersComponents

  class MyApplicationLoader extends ApplicationLoader {
    def load(context: Context) = {
      new MyComponents(context).application
    }
  }

  class MyComponents(context: Context) extends BuiltInComponentsFromContext(context) with HttpFiltersComponents {
    lazy val router = Router.empty
  }
//#basic

//#basicextended
  class MyApplicationLoaderWithInitialization extends ApplicationLoader {
    def load(context: Context) = {
      LoggerConfigurator(context.environment.classLoader).foreach {
        _.configure(context.environment, context.initialConfiguration, Map.empty)
      }
      new MyComponents(context).application
    }
  }
//#basicextended
}

package messages {
  // format: off
  import play.api._
  import play.api.routing.Router
  import play.api.ApplicationLoader.Context
  import play.filters.HttpFiltersComponents
  // format: on

//#messages
  import play.api.i18n._

  class MyComponents(context: Context)
      extends BuiltInComponentsFromContext(context)
      with I18nComponents
      with HttpFiltersComponents {
    lazy val router = Router.empty

    lazy val myComponent = new MyComponent(messagesApi)
  }

  class MyComponent(messages: MessagesApi) {
    // ...
  }
//#messages
}

package routers {
  import scalaguide.dependencyinjection.bar
  import scalaguide.dependencyinjection.controllers

  object router {
    type Routes = scalaguide.dependencyinjection.Routes
  }

//#routers
  import play.api._
  import play.api.ApplicationLoader.Context
  import play.filters.HttpFiltersComponents
  import router.Routes

  class MyApplicationLoader extends ApplicationLoader {
    def load(context: Context) = {
      new MyComponents(context).application
    }
  }

  class MyComponents(context: Context)
      extends BuiltInComponentsFromContext(context)
      with HttpFiltersComponents
      with controllers.AssetsComponents {
    lazy val barRoutes             = new bar.Routes(httpErrorHandler)
    lazy val applicationController = new controllers.Application(controllerComponents)

    lazy val router = new Routes(httpErrorHandler, applicationController, barRoutes, assets)
  }
//#routers
}

package controllers {
  import javax.inject.Inject

  import play.api.http.HttpErrorHandler
  import play.api.mvc._

  class Application @Inject() (cc: ControllerComponents) extends AbstractController(cc) {
    def index = Action(Ok)
    def foo   = Action(Ok)
  }

  trait AssetsComponents extends _root_.controllers.AssetsComponents {
    override lazy val assets = new scalaguide.dependencyinjection.controllers.Assets(httpErrorHandler, assetsMetadata)
  }

  class Assets(errorHandler: HttpErrorHandler, assetsMetadata: AssetsMetadata)
      extends _root_.controllers.Assets(errorHandler, assetsMetadata)
}

File: documentation/manual/working/javaGuide/main/dependencyinjection/code/javaguide/di/components/CompileTimeDependencyInjection.java
/*
 * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
 */
package javaguide.di.components;

// #basic-imports
import play.Application;
import play.ApplicationLoader;
import play.BuiltInComponentsFromContext;
import play.LoggerConfigurator;
import play.controllers.AssetsComponents;
import play.db.ConnectionPool;
import play.db.HikariCPComponents;
import play.filters.components.HttpFiltersComponents;
import play.mvc.Results;
import play.routing.Router;
import play.routing.RoutingDslComponentsFromContext;
// #basic-imports

import javaguide.dependencyinjection.controllers.Assets;
import javaguide.dependencyinjection.controllers.HomeController;

public class CompileTimeDependencyInjection {

  // #basic-app-loader
  public class MyApplicationLoader implements ApplicationLoader {

    @Override
    public Application load(Context context) {
      return new MyComponents(context).application();
    }
  }
  // #basic-app-loader

  // #basic-my-components
  public class MyComponents extends BuiltInComponentsFromContext implements HttpFiltersComponents {

    public MyComponents(ApplicationLoader.Context context) {
      super(context);
    }

    @Override
    public Router router() {
      return Router.empty();
    }
  }
  // #basic-my-components

  // #basic-logger-configurator
  // ###insert: import scala.compat.java8.OptionConverters;
  public class MyAppLoaderWithLoggerConfiguration implements ApplicationLoader {
    @Override
    public Application load(Context context) {

      LoggerConfigurator.apply(context.environment().classLoader())
          .ifPresent(
              loggerConfigurator ->
                  loggerConfigurator.configure(context.environment(), context.initialConfig()));

      return new MyComponents(context).application();
    }
  }
  // #basic-logger-configurator

  // #connection-pool
  public class MyComponentsWithDatabase extends BuiltInComponentsFromContext
      implements HikariCPComponents, HttpFiltersComponents {

    public MyComponentsWithDatabase(ApplicationLoader.Context context) {
      super(context);
    }

    @Override
    public Router router() {
      return Router.empty();
    }

    public SomeComponent someComponent() {
      // connectionPool method is provided by HikariCPComponents
      return new SomeComponent(connectionPool());
    }
  }
  // #connection-pool

  static class SomeComponent {
    SomeComponent(ConnectionPool pool) {
      // do nothing
    }
  }

  // #with-routing-dsl
  public class MyComponentsWithRouter extends RoutingDslComponentsFromContext
      implements HttpFiltersComponents {

    public MyComponentsWithRouter(ApplicationLoader.Context context) {
      super(context);
    }

    @Override
    public Router router() {
      // routingDsl method is provided by RoutingDslComponentsFromContext
      return routingDsl().GET("/path").routeTo(() -> Results.ok("The content")).build();
    }
  }
  // #with-routing-dsl

  // #with-generated-router
  public class MyComponentsWithGeneratedRouter extends BuiltInComponentsFromContext
      implements HttpFiltersComponents, AssetsComponents {

    public MyComponentsWithGeneratedRouter(ApplicationLoader.Context context) {
      super(context);
    }

    @Override
    public Router router() {
      HomeController homeController = new HomeController();
      Assets assets = new Assets(scalaHttpErrorHandler(), assetsMetadata());
      // ###replace: return new router.Routes(scalaHttpErrorHandler(), homeController,
      // assets).asJava();
      return new javaguide.dependencyinjection.Routes(
              scalaHttpErrorHandler(), homeController, assets)
          .asJava();
    }
  }
  // #with-generated-router
}
